src/ApplicationInsights.Kubernetes/Utilities/K8sDetectors/K8sDetectorFactory.cs

```diff
@@ -0,0 +1,13 @@
+namespace Microsoft.ApplicationInsights.Kubernetes.Utilities
+{
+    internal sealed class K8sDetectorFactory
+    {
+        private K8sDetectorFactory() { }
+
+        public static K8sDetectorFactory Instance { get; } = new K8sDetectorFactory();
+
+#pragma warning disable CA1822 // Mark members as static
+        public K8sDetector CreateDetector() => new K8sDetectorLinux(new K8sDetectorWindows(null));
```

Comment:
I think this is a bit too clever for what it really does :) 

(I mean the detector chain)

Reply:
Hmm, I don't really have a good reason to argue against the `too clever`. Do you really feel it becomes difficult to understand?

Reply:
Well, yes, see my comment above, where I did not notice that there is a concrete implementation of a method that calls an abstract "Imp" method.

So you created an abstract base class with two methods, plus two concrete classes, for something that boils down to 

```csharp
() => Directory.Exists("here") || Directory.Exists("there")
```

Reply:
:-) I guess in this instance, it will be a longshot for me to say that is for the extensibility (OC principle). It sometimes is really difficult to draw the line.

Despite this PR, how to do determine when to introduce some trick to make the code extensible vs when to write straightforward code and then expect no change happens?

Reply:
Agree this is not always a straightforward call. I would consider the following:
  - How likely it is that we will have to introduce more flexible implementation in future?
  - If we have to change the implementation, how costly will the change be?

In this this case the i simplest interface is a `Func<Bool>`, the implementation is a one-liner, and it is not very likely to change in future, so I am inclined to go with the simple thing.